library/kani_core/src/mem.rs

```diff
@@ -123,9 +123,17 @@ macro_rules! kani_mem {
             <T as Pointee>::Metadata: PtrProperties<T>,
         {
             let (thin_ptr, metadata) = ptr.to_raw_parts();
+            // Need to assert `is_initialized` because non-determinism is used under the hood, so it
+            // does not make sense to use it inside assumption context.
             metadata.is_ptr_aligned(thin_ptr, Internal)
                 && is_inbounds(&metadata, thin_ptr)
-                && is_initialized(ptr, 1)
+                && {
+                    assert!(
```

Comment:
Neat!

Wouldn't this still result in using `is_initialized` in an assumption though since we encode `assert!` as assert followed by assume?

Reply:
As far as I understand, using it in an assumption context it would be a no-op, wouldn't it?

Reply:
Do we have an API that generates assert only? This would be easy to implement if needed.

Reply:
I don't think we have one currently. We can add one (e.g. `check`) that is codegened as an assert only (`__CPROVER_assert`).

Reply:
I fixed it in the latest commit, also made sure that the instrumentation adds assertions only